lib/db_connection.ex

```diff
@@ -738,45 +736,134 @@ defmodule DBConnection do
-        raise DBConnection.ConnectionError, "connection is closed"
+      {:error, err, _} ->
+        raise err
     end
   end
   def run(pool, fun, opts) do
+    conn = checkout!(pool, opts)
+    try do
+      fun.(conn)
+    after
+      checkin(conn, opts)
+    end
+  end
+  @doc """
+  Acquire a lock on a connection.
+
+  Returns `{:ok, conn}` on success or `{;error, exception}` if there is an
+  error.
+
+  To use the locked connection call requests with the connection
+  reference returned. If the connection disconnects all future calls using that
+  connection reference will fail.
+
+  Use `checkin/2`, `commit_checkin/2` or `rollback_checkin/2` to release the
+  lock and return the connection to the pool.
+
+  ### Options
+
+    * `:pool_timeout` - The maximum time to wait for a reply when making a
+    synchronous call to the pool (default: `5_000`)
+    * `:queue` - Whether to block waiting in an internal queue for the
+    connection's state (boolean, default: `true`)
+    * `:timeout` - The maximum time that the caller is allowed the
+    to hold the connection's state (default: `15_000`)
+
+  The pool may support other options.
+
+  ### Example
+
+      {:ok, conn} = DBConnection.checkout(pool)
+      try do
+        DBConnection.execute(conn, "SELECT * FROM table", [])
+      after
+        DBConnection.checkin(conn)
+      end
+  """
+  @spec checkout(pool :: GenServer.t, opts :: Keyword.t) ::
+    {:ok, t} | {:error, Exception.t}
+  def checkout(pool, opts \\ []) do
     case checkout(pool, nil, opts) do
-      {:ok, conn, _, _} ->
-        try do
-          fun.(conn)
-        after
-          checkin(conn, opts)
-        end
-      {:error, err, _} ->
+      {:ok, conn, _conn_state, _meter} ->
+        {:ok, conn}
+      {:error, err, _meter} ->
+        {:error, err}
+      {kind, reason, stack, _meter} ->
+        :erlang.raise(kind, reason, stack)
+    end
+  end
+
+  @doc """
+  Acquire a lock on a connection.
+
+  Returns the connection reference on success, and raises an exception on error.
+  See `checkout/2`.
+  """
+  @spec checkout!(pool :: GenServer.t, opts :: Keyword.t) :: t
+  def checkout!(pool, opts \\ []) do
+    case checkout(pool, nil, opts) do
+      {:ok, conn, _conn_state, _meter} ->
+        conn
+      {:error, err, _meter} ->
         raise err
-      {kind, reason, stack, _} ->
+      {kind, reason, stack, _meter} ->
         :erlang.raise(kind, reason, stack)
     end
   end
 
+
   @doc """
-  Acquire a lock on a connection and run a series of requests inside a
-  transaction. The result of the transaction fun is return inside an `:ok`
-  tuple: `{:ok, result}`.
+  Release a lock on a connection.
 
-  To use the locked connection call the request with the connection
-  reference passed as the single argument to the `fun`. If the
-  connection disconnects all future calls using that connection
-  reference will fail.
+  Returns `:ok` on success, otherwise `{:error, exception}` on error.
```

Comment:
What happens if the connection was already checked in? Or if it was disconnected between the last command and checking in?